File: 438.findAllAnagramsInaString/findAllAnagramsInastring.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> res;
        int s_len = s.length();
        int p_len = p.length();
        if (s_len < p_len)
            return res;
        int hash[26] = {0};
        for (int i = 0; i < p_len; i++)
            hash[p[i] - 'a']++;
        int count = p_len;
        int left = 0, right = 0;
        while (right < s_len) {
            if (--hash[s[right++] - 'a'] >= 0)
                count--;
            if (count == 0)
                res.push_back(left);
            if (right - left == p_len) {
                if (++hash[s[left++] - 'a'] > 0)
                    count++;
            }
        }
        return res;
    }
};

int main()
{
    Solution s;
    string ss = "abab";
    string p = "ab";
    vector<int> ans = s.findAnagrams(ss, p);
    vector<int>::iterator it = ans.begin();
    while (it != ans.end()) {
        cout << *it << " ";
        ++it;
    }
    cout << endl;
    return 0;
}

File: 438.findAllAnagramsInaString/findAllAnagramsInastring.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        int[] cntP = new int[26];
        for (int i = 0; i < p.length(); i++) cntP[p.charAt(i) - 'a']++;

        int[] cntS = new int[26];
        for (int i = 0; i <= s.length() - p.length(); i++) {
            if (i == 0) {
                for (int j = 0; j < p.length(); j++) {
                    cntS[s.charAt(j) - 'a']++;
                }
            } else {
                cntS[s.charAt(i - 1) - 'a']--;
                cntS[s.charAt(i + p.length() - 1) - 'a']++;
            }
            
            if (Arrays.equals(cntS, cntP)) res.add(i);
        }
        return res;
    }
}